lib/elixir/lib/module/types/descr.ex

```diff
@@ -954,6 +981,197 @@ defmodule Module.Types.Descr do
+      :open -> {:{}, [], tuple_fields_to_quoted(tag, fields) ++ [{:..., [], nil}]}
+    end
+  end
+
+  defp tuple_fields_to_quoted(tag, map) do
+    sorted = Enum.sort(map)
+
+    for {_, type} <- sorted,
+        not (tag == :open and optional?(type) and term_type?(type)) do
+      cond do
+        not optional?(type) -> to_quoted(type)
+        empty?(type) -> {:not_set, [], []}
+        true -> {:if_set, [], [to_quoted(type)]}
+      end
+    end
+  end
+
+  # Emptiness checking for tuples.
+  #
+  # Short-circuits if it finds a non-empty map tuple in the union.
+  # Since the algorithm is recursive, we implement the short-circuiting
+  # as throw/catch.
+  #
+  # For tuples, the explicit types do not contain optional
+  defp tuple_empty?(dnf) do
+    try do
+      for {tag, pos, negs} <- dnf do
+        tuple_empty?(tag, pos, negs)
+      end
+
+      true
+    catch
+      :not_empty -> false
+    end
+  end
+
+  defp tuple_empty?(_, _, []), do: throw(:not_empty)
+  defp tuple_empty?(_, _, [{:open, fs} | _]) when fs == %{}, do: true
+
+  defp tuple_empty?(tag, fields, [{neg_tag, neg_fields} | negs]) do
+    try do
+      # Indices that exists in the negative tuple, but not in the positive one
+      for {neg_index, neg_type} <- neg_fields, not is_map_key(fields, neg_index) do
+        cond do
+          # This index is not in the closed positive tuple, so there is no value in common
+          # Tthere are no explicit optional types in tuples
+          tag == :closed ->
+            throw(:discard_negative)
+
+          # There may be value in common
+          tag == :open ->
+            diff = difference(term_or_optional(), neg_type)
+
+            # Remove all the elements with an index larger or equal to neg_index
+            keys_to_drop = for key <- Map.keys(fields), key >= neg_index, do: key
+            closed_pos = Map.drop(fields, keys_to_drop)
+
+            # For neg_key index, remove the absent part
+            open_pos = Map.put(fields, neg_index, remove_optional(diff))
+
+            tuple_empty?(:closed, closed_pos, negs) and tuple_empty?(tag, open_pos, negs)
+        end
+      end
+
+      # Indices from the positive tuple that may be present in the negative one
+      for {index, type} <- fields do
+        case neg_fields do
+          %{^index => neg_type} ->
+            diff = difference(type, neg_type)
+            empty?(diff) or tuple_empty?(tag, Map.put(fields, index, diff), negs)
+
+          %{} ->
+            if neg_tag == :closed do
+              throw(:discard_negative)
+            else
+              # an absent key in a open negative map can be ignored
+              diff = difference(type, map_tag_to_type(neg_tag))
+              empty?(diff) or tuple_empty?(tag, Map.put(fields, index, diff), negs)
+            end
+        end
+      end
+
+      true
+    catch
+      :discard_negative -> tuple_empty?(tag, fields, negs)
+    end
+  end
```

Comment:
Can we rewrite it to a non-raising versaion like the map one is currently implemented? :)

Reply:
@gldubc FYI https://github.com/elixir-lang/elixir/pull/13560

Reply:
Yes! I see this now :)